cpp/include/cudf/utilities/span.hpp

```diff
@@ -87,12 +93,40 @@ template <typename T, std::size_t Extent = dynamic_extent>
 struct host_span : public span_base<T, Extent, host_span<T, Extent>> {
   using base = cudf::detail::span_base<T, Extent, host_span<T, Extent>>;
   using base::base;
+
+  template <typename Container>
+  static bool constexpr is_supported_container()
+  {
+    return std::is_same<std::vector<T>, Container>() ||
+           std::is_same<thrust::host_vector<T>, Container>();
+  }
+
+  // hacky way of implementing span::span(R&&) for ranges;
+  template <typename Container, std::enable_if_t<is_supported_container<Container>()>* = nullptr>
+  explicit host_span(Container source) : base(source.data(), source.end())
+  {
+  }
 };
 
 template <typename T, std::size_t Extent = dynamic_extent>
 struct device_span : public span_base<T, Extent, device_span<T, Extent>> {
   using base = cudf::detail::span_base<T, Extent, device_span<T, Extent>>;
   using base::base;
+
+  template <typename Container>
+  static bool constexpr is_supported_container()
```

Comment:
I'd go with `is_supported_host/device_container` for greater clarity.

Reply:
Was thinking it was redundant since it's contained within the class. I'll change it.

Reply:
`thrust::device_vector::data()` returns `thrust::device_ptr`, so the template isn't correct. Going to get rid of `is_supported_container` entirely and just make individual explicit constructors.